Counter/src/app/shared/hall-of-fame-service.spec.ts

```diff
@@ -0,0 +1,188 @@
+import { HallOfFameService } from "./hall-of-fame-service"
+import { Pokemon } from "./pokemon";
+import { TestBed } from '@angular/core/testing';
+import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';
+import { UserService } from "./user-service";
+import { map, Observable, of, BehaviorSubject, take } from "rxjs";
+
+describe('HallOfFameService', () => {
+  let service: HallOfFameService;
+  
+  let mockHttp: HttpTestingController;
+
+  let userServiceStub: Partial<UserService> = {
+    getId(): Observable<number> {
+      return of(1);
+    }
+  };
+
+  beforeEach(() => {
+    TestBed.configureTestingModule({
+        imports: [HttpClientTestingModule],
+        providers: [ { provide: UserService, useValue: userServiceStub } ]
+    });
+    service = TestBed.inject(HallOfFameService);
+    mockHttp = TestBed.inject(HttpTestingController);
+  });
+
+  it('should be created', () => {
+    expect(service).toBeTruthy();
+  });
+
+  describe('getShiny', () => {
+    it('should return shiny', () => {
+      const testData = [{id: 1, name: "Raticate", count: 3, userId: 1},
+      {id: 2, name: "Audino", count: 4914, userId: 1},
+      {id: 3, name: "Cresselia", count: 3629, userId: 1},
+      {id: 4, name: "Axew", count: 497, userId: 1}];
+
+      service.getShiny().subscribe((result: Pokemon[]) => {
+        expect(result).toEqual(testData);
+        expect(result.length).toBe(4);
+      })
+
+      const req = mockHttp.expectOne('api/shiny');
+      expect(req.request.url).toEqual('api/shiny');
+      expect(req.request.method).toBe('GET');
+
+      req.flush(testData);
+    }),
+
+    describe('should return different pokemon depending on userId', () => {
+      let idStream = new BehaviorSubject(0);
```

Comment:
You have to be careful with test data and the scope it has within a test. You don't want to create problems when multiple tests are running on (and changing) the same data. Test data should be reset after every test. To do this, you use a `beforeEach(...)` block. This means before each test it will run this method. This is where you can set up your data for the current `describe`.

It is good practice to limit your test data to the minimum amount you need for the test to pass. When testing list related things you only need 2 records unless you have a specific scenario you're trying to test with a larger set.

In that test data you also want to avoid values that make sense to your application. So instead of "Axew" you might use "Unit Test Pokemon 1". You can often find bugs in your code when you send in data that is out of the norm. This goes for all string values you have in your tests. Numbers I typically do things like 111, 222, 333, etc.